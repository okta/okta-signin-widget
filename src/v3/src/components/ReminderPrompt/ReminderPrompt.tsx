/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Alert, Box, Link } from '@okta/odyssey-react-mui';
import { HTMLReactParserOptions } from 'html-react-parser';
import { h } from 'preact';
import { useEffect, useRef, useState } from 'preact/hooks';

import { useWidgetContext } from '../../contexts';
import { useHtmlContentParser, useOnSubmit } from '../../hooks';
import { ReminderElement, UISchemaElementComponent } from '../../types';
import { getLinkReplacerFn, SessionStorage } from '../../util';
import TextWithActionLink from '../TextWithActionLink';

export const DEFAULT_TIMEOUT_MS = 30_000;

const ReminderPrompt: UISchemaElementComponent<{
  uischema: ReminderElement
}> = ({ uischema }) => {
  const {
    loading,
  } = useWidgetContext();
  const {
    content,
    step,
    actionParams,
    isActionStep,
    buttonText,
    timeout: customTimeout,
    contentClassname,
    contentHasHtml,
  } = uischema.options;
  const onSubmitHandler = useOnSubmit();
  const parsedContent = useHtmlContentParser(content, uischema.parserOptions);

  const [show, setShow] = useState<boolean>(false);
  const timerRef = useRef<number | undefined>();

  const startTimer = () => {
    setShow(false);
    const timeStamp = SessionStorage.getResendTimestamp();
    if (!timeStamp) {
      SessionStorage.setResendTimestamp(Date.now().toString());
    }
    if (timerRef) {
      window.clearInterval(timerRef.current);
    }

    timerRef.current = window.setInterval(() => {
      const ts = SessionStorage.getResendTimestamp() || Date.now().toString();
      const start = parseInt(ts, 10);
      const now = Date.now();
      const timeout = typeof customTimeout === 'number' ? customTimeout : DEFAULT_TIMEOUT_MS;
      if (now - start >= timeout) {
        setShow(true);
        window.clearInterval(timerRef.current);
        SessionStorage.removeResendTimestamp();
      }
    }, 250);
  };

  useEffect(() => {
    startTimer();

    return () => {
      window.clearInterval(timerRef.current);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const resendHandler = async () => {
    if (loading) {
      return;
    }
    startTimer();

    await onSubmitHandler({ step, isActionStep, params: actionParams });
  };

  const renderActionLink = () => {
    if (!buttonText) {
      return undefined;
    }

    return (
      // eslint-disable-next-line jsx-a11y/anchor-is-valid
      <Link
        // eslint-disable-next-line no-script-url
        href="javascript:void(0);"
        variant="monochrome"
        onClick={() => resendHandler()}
      >
        {buttonText}
      </Link>
    );
  };

  const renderAlertContent = () => {
    if (contentHasHtml && contentClassname) {
      const parserOptions: HTMLReactParserOptions = {};
      parserOptions.replace = getLinkReplacerFn(parserOptions, 'monochrome');
      return (
        <TextWithActionLink
          uischema={{
            type: 'TextWithActionLink',
            parserOptions,
            options: {
              contentClassname,
              content,
              step,
              isActionStep,
              actionParams,
            },
          }}
        />
      );
    }
    return (
      <Box marginBlockEnd={2}>{parsedContent}</Box>
    );
  };

  return show ? (
    <Box marginBlockEnd={4}>
      <Alert
        severity="warning"
        variant="infobox"
        sx={{
          // TODO: OKTA-534606 - switch to ODS component which has this fix
          '& .MuiAlert-message': {
            overflow: 'visible',
          },
        }}
      >
        {renderAlertContent()}
        {renderActionLink()}
      </Alert>
    </Box>
  ) : null;
};

export default ReminderPrompt;
