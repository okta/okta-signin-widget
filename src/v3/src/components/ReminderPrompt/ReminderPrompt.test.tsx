/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import '@testing-library/jest-dom';

import {
  cleanup, fireEvent, render, within,
} from '@testing-library/preact';
import { h } from 'preact';
import { act } from 'preact/test-utils';

import {
  ReminderElement,
  UISchemaElementComponentProps,
} from '../../types';
import ReminderPrompt, { DEFAULT_TIMEOUT_MS } from './ReminderPrompt';

// @ts-expect-error Expected 0 arguments, but got 1
jest.useFakeTimers('modern');

const mockSubmitHook = jest.fn().mockImplementation(() => ({}));
jest.mock('../../hooks', () => ({
  useOnSubmit: () => mockSubmitHook,
}));

jest.mock('../../contexts', () => ({
  useWidgetContext: () => ({
    loading: jest.fn().mockReturnValue(false)(),
  }),
}));

describe('ReminderPrompt', () => {
  beforeEach(() => {
    mockSubmitHook.mockRestore();
  });

  afterEach(() => {
    cleanup();
  });

  it('should show prompt with working link after default timeout passes', async () => {
    const step = 'currentAuthenticator-resend';
    const props: UISchemaElementComponentProps & { uischema: ReminderElement; } = {
      uischema: {
        type: 'Reminder',
        options: {
          content: 'Didnt receive the email?',
          buttonText: 'Send again?',
          step,
          actionParams: { resend: true },
          isActionStep: true,
        },
      },
    };

    const { container, getByRole } = render(<ReminderPrompt {...props} />);

    expect(container.firstChild).toBeNull();
    act(() => {
      // @ts-expect-error setSystemTime does not exist on type 'typeof jest'
      jest.setSystemTime(Date.now() + DEFAULT_TIMEOUT_MS);
      jest.runOnlyPendingTimers();
    });

    expect(container.firstChild).not.toBeNull();

    const box = getByRole('alert');
    const sendAgainLink = await within(box).findByText('Send again?');

    expect(container).toMatchSnapshot();

    expect(mockSubmitHook).not.toHaveBeenCalled();
    fireEvent.click(sendAgainLink);
    expect(mockSubmitHook).toHaveBeenCalledTimes(1);
    expect(mockSubmitHook).toHaveBeenCalledWith({
      step,
      isActionStep: true,
      params: { resend: true },
    });
  });

  it('should show prompt with working link after default timeout passes where ctaText contains HTML', async () => {
    const step = 'currentAuthenticator-resend';
    const props: UISchemaElementComponentProps & { uischema: ReminderElement; } = {
      uischema: {
        type: 'Reminder',
        options: {
          content: "Didn't receive the email? Click <a href='#' class='send-again'>send again</a>",
          contentHasHtml: true,
          step,
          contentClassname: 'send-again',
          actionParams: { resend: true },
          isActionStep: true,
        },
      },
    };

    const { container, getByRole } = render(<ReminderPrompt {...props} />);

    expect(container.firstChild).toBeNull();

    act(() => {
      // @ts-expect-error setSystemTime does not exist on type 'typeof jest'
      jest.setSystemTime(Date.now() + DEFAULT_TIMEOUT_MS);
      jest.runOnlyPendingTimers();
    });

    expect(container.firstChild).not.toBeNull();

    const box = getByRole('alert');
    const sendAgainLink = await within(box).findByText('send again');

    expect(container).toMatchSnapshot();

    expect(mockSubmitHook).not.toHaveBeenCalled();
    fireEvent.click(sendAgainLink);
    expect(mockSubmitHook).toHaveBeenCalledTimes(1);
    expect(mockSubmitHook).toHaveBeenCalledWith({
      step,
      isActionStep: true,
      params: { resend: true },
    });
  });

  it('should show prompt after custom timeout passes', async () => {
    const CUSTOM_TIMEOUT = 1_000 * 60 * 5;

    const props: UISchemaElementComponentProps & { uischema: ReminderElement; } = {
      uischema: {
        type: 'Reminder',
        options: {
          content: 'Didnt receive the email?',
          buttonText: 'Send again?',
          timeout: CUSTOM_TIMEOUT,
          step: 'currentAuthenticator-resend',
          actionParams: { resend: true },
          isActionStep: true,
        },
      },
    };

    const { container } = render(<ReminderPrompt {...props} />);

    expect(container).toMatchSnapshot();

    act(() => {
      // @ts-expect-error setSystemTime does not exist on type 'typeof jest'
      jest.setSystemTime(Date.now() + CUSTOM_TIMEOUT);
      jest.runOnlyPendingTimers();
    });

    expect(container.firstChild).not.toBeNull();
  });
});
