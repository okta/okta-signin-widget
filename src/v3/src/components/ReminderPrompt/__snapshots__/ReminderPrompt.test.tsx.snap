// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReminderPrompt should show prompt after custom timeout passes 1`] = `<div />`;

exports[`ReminderPrompt should show prompt with working link after default timeout passes 1`] = `
<div>
  <div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-calloutWarning MuiAlert-callout emotion-1"
      role="alert"
    >
      <div
        class="MuiAlert-icon emotion-2"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-3"
          data-testid="ReportProblemOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
          />
        </svg>
      </div>
      <div
        class="MuiAlert-message emotion-4"
      >
        <span
          class="MuiBox-root emotion-5"
        >
          warning
        </span>
        <div
          class="MuiBox-root emotion-6"
        >
          <div
            class="MuiBox-root emotion-7"
          >
            Didnt receive the email?
          </div>
          <a
            class="MuiTypography-root MuiTypography-monochrome MuiLink-root MuiLink-underlineAlways emotion-8"
            href="javascript:void(0);"
          >
            Send again?
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ReminderPrompt should show prompt with working link after default timeout passes where ctaText contains HTML 1`] = `
<div>
  <div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-calloutWarning MuiAlert-callout emotion-1"
      role="alert"
    >
      <div
        class="MuiAlert-icon emotion-2"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-3"
          data-testid="ReportProblemOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
          />
        </svg>
      </div>
      <div
        class="MuiAlert-message emotion-4"
      >
        <span
          class="MuiBox-root emotion-5"
        >
          warning
        </span>
        <div
          class="MuiBox-root emotion-6"
        >
          <div
            class="MuiBox-root emotion-6"
          >
            Didn't receive the email? Click 
            <a
              class="MuiTypography-root MuiTypography-monochrome MuiLink-root MuiLink-underlineAlways send-again emotion-8"
              href="#"
            >
              send again
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
